libucontext: Add package

VolkerChristian opened the pull request:
Maintainer: @VolkerChristian

Description (from libucontext github page):
libucontext ([https://github.com/kaniini/libucontext](https://github.com/kaniini/libucontext)) is a library which provides the ucontext.h C syscall API. Unlike other implementations, it faithfully follows the kernel process ABI when doing context swaps. libucontext is used on almost all musl distributions to provide the legacy ucontext.h API.

This is a development package only. No need to install a package on the router. It is used to link libraries/applications against it which need the system calls
- makecontext
- swapcontext
- getcontext
- setcontext

E.g. the asynchronous API of libmariadb (c-connector) uses this calls. Because libmusl didn't provide that system calls this asynchronous API is currently (without libucontexe) not working - it segfaults. A PR addressing libmariadb is ready and will be send as soon as libucontext has been accepted and merged.

Compile tested on:
- aarch64_cortex-a53
- aarch64_cortex-a72
- aarch64_generic
- arm_arm926ej-s
- arm_cortex-a15_neon-vfpv4
- arm_cortex-a7
- arm_cortex-a7_neon-vfpv4
- arm_cortex-a7_vfpv4
- arm_cortex-a9
- arm_cortex-a9_neon
- arm_cortex-a9_vfpv3-d16
- arm_fa526
- arm_mpcore
- arm_xscale
- i386_pentium-mmx
- mips_24kc
- mips64_octeonplus
- mipsel_24kc
- mipsel_24kc_24kf
- mipsel_74kc
- mipsel_mips32
- mips_mips32
- powerpc_464fp
- powerpc_8548
- riscv64_riscv64

Run tested:
- GL.iNet GL-A1300 (arm_cortex-a7_neon-vfpv4)
- GL.iNet GL-MT3000 (aarch64_cortex-a53)
- Linksys MR8300 (arm_cortex-a7_neon_vfpv4)
- Netgear R7800 (arm_cortex-a15_neon_vfpv4)
- TP-Link Archer A7v5 (mipsel_24kc)

1715173329 commented:
https://github.com/openwrt/packages/pull/21924#issuecomment-1694231327

i would still suggest using meson build and PKG_BUILD_FLAGS.

VolkerChristian commented:
> [#21924 (comment)](https://github.com/openwrt/packages/pull/21924#issuecomment-1694231327)
> 
> i would still suggest using meson build and PKG_BUILD_FLAGS.

Thank you for this info! I am sorry, I have overlooked your first comment about this package.
I looked at your libucontext/Makefile and was pleased to see that it is very similar to mine.
Out of curiosity: why didn't you create a pull request?

I will follow your hint using meson and also the flag concerning the mips16 architecture!

Thanks!

1715173329 commented:
> Out of curiosity: why didn't you create a pull request?

I ported it as a part of gcompat[^1][^2] support. Unfortunately OpenWrt's GCC toolchain seems to have some issues (broken `-static-pie`) and I'm not pro enough to resolve it (xD
So all of these packages are actually untested.

> Just another question: Is it preferred to use MAKE_FLAGS and MAKE_INSTALL_FLAGS over "define Build/Compile" and "define Build/Installed" like in the commit named "libucontext: Use MAKE_FLAGS and MAKE_INSTALL_FLAGS instead of..."?

Basically yes. `MAKE_*_FLAGS` helps to simplify the Makefile.

btw, i guess we can patch https://github.com/kaniini/libucontext/blob/be80075e957c4a61a6415c280802fea9001201a2/meson.build#L11 to `target_machine.cpu_family()` to avoid `UC_ARCH`.

[^1]: https://git.adelielinux.org/adelie/gcompat
[^2]: https://github.com/1715173329/packages-immortalwrt/commits/gcompat

VolkerChristian commented:
Hi,

I have switched to meson now. All seems fine. I tested compilation on many architectures and the $(MESON_ARCH) was always right! Great! Thanks for the hint about meson!

I left the previous code as a comment in the Makefile - will remove it if all is fine.

> > Out of curiosity: why didn't you create a pull request?
> 
> I ported it as a part of gcompat[1](#user-content-fn-1-d512c86222b53e78a6d794fe0cb243f7)[2](#user-content-fn-2-d512c86222b53e78a6d794fe0cb243f7) support. Unfortunately OpenWrt's GCC toolchain seems to have some issues (broken `-static-pie`) and I'm not pro enough to resolve it (xD So all of these packages are actually untested.
> 
Hm, I am also not a pro about OpenWrt - this is my first contribution and I am still learning ...
Anyway, as soon as I have time I'll see if I can help.

> ## Footnotes
> 1. https://git.adelielinux.org/adelie/gcompat [↩](#user-content-fnref-1-d512c86222b53e78a6d794fe0cb243f7)
> 2. https://github.com/1715173329/packages-immortalwrt/commits/gcompat [↩](#user-content-fnref-2-d512c86222b53e78a6d794fe0cb243f7)

1715173329 commented:
> I tested compilation on many architectures and the $(MESON_ARCH) was always right!

Nah, I meant use meson for build. So you can drop `MAKE_FLAGS` and `Build/Compile` etc.

`meson.build` doesn't include examples, I'm not sure if they're necessary or important. For me it's not: why would we use test/example binraries on a router? Shipping dynamic libraries should be enough.

It would be nice to copy pkgconfig, headers and dynamic libraries at `InstallDev` stage, so other programs can find it and link to it correctly:

```
define Build/InstallDev
	$(INSTALL_DIR) $(1)/usr/include $(1)/usr/lib/pkgconfig
	$(CP) $(PKG_INSTALL_DIR)/usr/include/libucontext $(1)/usr/include/
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libucontext* $(1)/usr/lib/
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libucontext.pc $(1)/usr/lib/pkgconfig/
endef

define Package/libucontext/install
	$(INSTALL_DIR) $(1)/usr/lib
	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libucontext*.so* $(1)/usr/lib/
endef
```

> Anyway, as soon as I have time I'll see if I can help.

That would be great!

Thank you.

VolkerChristian commented:
> > I tested compilation on many architectures and the $(MESON_ARCH) was always right!
> 
> Nah, I meant use meson for build. So you can drop `MAKE_FLAGS` and `Build/Compile` etc.

Arg, my fault - i still have called `Build/Compile/Default` instead `Build/Compile/Meson`. I changed this which makes MAKE_FLAGS obsolet now. Fine! :-)
> 
> `meson.build` doesn't include examples, I'm not sure if they're necessary or important. For me it's not: why would we use test/example binraries on a router? Shipping dynamic libraries should be enough.

Yes, examples still where compilable due to the call to `Build/Compile/Default` - now it's not anymore. Thus i have removed it.

> 
> It would be nice to copy pkgconfig, headers and dynamic libraries at `InstallDev` stage, so other programs can find it and link to it correctly:
> 
> ```
> define Build/InstallDev
> 	$(INSTALL_DIR) $(1)/usr/include $(1)/usr/lib/pkgconfig
> 	$(CP) $(PKG_INSTALL_DIR)/usr/include/libucontext $(1)/usr/include/
> 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libucontext* $(1)/usr/lib/
> 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libucontext.pc $(1)/usr/lib/pkgconfig/
> endef
> 
> define Package/libucontext/install
> 	$(INSTALL_DIR) $(1)/usr/lib
> 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libucontext*.so* $(1)/usr/lib/
> endef
> ```

Done!  :-)

1715173329 commented on libs/libucontext/Makefile:
```diff
@@ -0,0 +1,102 @@
+PKG_NAME:=libucontext
+PKG_VERSION:=1.2
+PKG_RELEASE:=1
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).zip
+PKG_SOURCE_URL:=https://codeload.github.com/kaniini/libucontext/zip/v$(PKG_VERSION)?
+PKG_HASH:=77878849d938bfac041702b38e34aac2c09e90b3c94d13388c751b4e434ddc79
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
```
```suggestion

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://codeload.github.com/kaniini/libucontext/tar.gz/v$(PKG_VERSION)?
PKG_HASH:=2657e087c493263e7bbbde152a5bc08ce22dc5a7970887ac4fd251b90b58401f
```

it's recommended to use tar.gz file.
setting `PKG_BUILD_DIR` is not necessary as it's default

1715173329 commented on libs/libucontext/Makefile:
```diff
@@ -0,0 +1,102 @@
+PKG_LICENSE:=ISC
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
```
```suggestion
PKG_BUILD_FLAGS:=no-mips16
PKG_BUILD_PARALLEL:=1
```

1715173329 commented on libs/libucontext/Makefile:
```diff
@@ -0,0 +1,102 @@
+#endif
+
+# Do not use mips16 as the assempler code of libucontext did not support it
+ifdef CONFIG_USE_MIPS16
+  TARGET_CFLAGS+=-mno-mips16 -mno-interlink-mips16
+endif
```
```suggestion
```
use PKG_BUILD_FLAGS instead

1715173329 commented on libs/libucontext/Makefile:
```diff
@@ -0,0 +1,102 @@
+  libraries/applications which need the SYS-V ucontext API.
+endef
+
+define Build/Compile
+	$(call Build/Compile/Meson)
+	$(call Build/Compile/Meson,test_libucontext_posix)
+	$(call Build/Compile/Meson,test_libucontext)
+endef
```
```suggestion
```

now we can remove this. it's default behavior

VolkerChristian commented:
I have now integrated all your suggestions. It is now more or less your pull request. ;-) I will add you as co-author when I  squash all commits into on! I really appreciate your detailed review - I learned a lot! Thank you.

Additional libucontext-tests now depends on libucontext and did not install the libraries itself to avoid errors during installation if both packages should be installed. And the last modification concerns the installed shared libraries. Only the libucontext*.so.1 libraries should be installed. There is no need for the non-versioned libraries on the device at runtime.

What do you think - shall I squash the commits for merge?

1715173329 commented:
> What do you think - shall I squash the commits for mergeing?

Yep, I'm fine to merge it. But let's address the last comment here: https://github.com/openwrt/packages/pull/21933#discussion_r1312887617

VolkerChristian commented:
> > What do you think - shall I squash the commits for mergeing?
> 
> Yep, I'm fine to merge it. But let's address the last comment here: [#21933 (comment)](https://github.com/openwrt/packages/pull/21933#discussion_r1312887617)

Oh, i missed that! :-(

Please modify the commit message to suit your data in the Co-authored-by: Trailer if not correct!

Next pull request will address libmariadb to link against libucontext - I hope, this will not be that will be more strait forward than this one. But I have to modify my current modifications reflecting the fact that libucontext is only build on musl platforms. The current modification assume that libucontext is available everywhere ...

Thank you and have a nice day!

1715173329 commented on libs/libucontext/Makefile:
```diff
@@ -0,0 +1,61 @@
+PKG_NAME:=libucontext
+PKG_VERSION:=1.2
+PKG_RELEASE:=1
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).zip
```
```suggestion

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).zip
```

i'm okay if you still prefer zip, anyway please add a line break here.

VolkerChristian replied:
No I don't prefere zip :-(
It seems i am a little bit unconcentrated and also confused not missing any hints from you :-( Sorry!

1715173329 commented:
> libucontext is only build on musl platforms.

Doesn't glibc provide these ABI already?

libucontext's README says `ucontext implementation featuring glibc-compatible ABI`, does I misunderstand anything?

VolkerChristian commented:
> > libucontext is only build on musl platforms.
> 
> Doesn't glibc provide these ABI already?

Yes, glibc already provides it. But i need to patch the CMakeLists.txt file to link against libucontext on musl and not link to anything else on glibc platforms.
Currently my mod links always to libucontext ... 

> 
> libucontext's README says `ucontext implementation featuring glibc-compatible ABI`, does I misunderstand anything?
> 
> > Please modify the commit message
> 
> 1. Please wrap a new line for every 70 chars
> 2. Signoff block should be:
>    ```
>    Co-developed-by: Tianling Shen <cnsztl@immortalwrt.org>
>    Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
>    Signed-off-by: Volker Christian <me@vchrist.at>
>    ```

Modified!

1715173329 commented:
> Modified!

I guess you fotgot to wrap commit message ;)

VolkerChristian commented:
> > Modified!
> 
> I guess you fotgot to wrap commit message ;)

Shame on me! :-(

1715173329 reviewed (approved):
LGTM, thank you!

VolkerChristian commented:
> LGTM, thank you!

Thank you!
